cores/esp8266/core_esp8266_si2c.cpp

```diff
@@ -262,18 +218,18 @@ void Twi::enableSlave()
     }
 }
 
-void ICACHE_RAM_ATTR Twi::busywait(unsigned char v)
+ void ICACHE_RAM_ATTR Twi::busywait(unsigned int v)
 {
     unsigned int i;
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
-    unsigned int reg;
-    for (i = 0; i < v; i++)
+    for (i = 0; i < v; i++)  // loop time is 5 machine cycles: 31.25ns @ 160MHz, 62.5ns @ 80MHz
     {
-        reg = GPI;
+		asm("nop"); // minimum element to keep GCC from optimizing this function out.
```

Comment:
This will drift when there are interrupts happening in the middle of the loop. The effective wait doesn't take into account the time spent in the ISRs.
That is no different than what is happening in tje current code, but it is a potential point of failure. 
We discussed the polledTimeout approach, which is commented out below. I don't remember the conclusion about it anymore, was it just a bit slower, i. e. lower max clock speed?

Reply:
Quite a bit slower; the minimum granularity of oneShotFastNs was around 200ns, which limits the top speed to 379KHz @80 and 658KHz @160.  Most people are using either 100KHz or 400KHz, very few faster or slower than that.  If you look at one of the earlier commits, I'd tried it and forgot and left parts of it commented out in the commit.  If an interrupt hits, the actual clock period will be longer with a fixed busywait, but it's a static bus (purely edge-driven) so the slave will ride it out.  Best guess is interrupts are only 2-5% of the time even with busy WiFi traffic.  I clocked the WiFi services at anywhere from low microseconds to 2.5 milliseconds range.  Low microseconds won't effect the 5us period of 100KHz much, but the longer ones will.  

Hmmm.  Side issue for later: those occasional 2.5ms WiFi services could cause the I2C slave to timeout and abort, as it has a 10 millisecond timeout.  I don't know if ets_timer_arm_new() is higher priority than the WiFi events.

Here's the difference due to the granularity for the most common clock speeds:

requested bus clock = 50000 (currently needed for the ESP slave)
actual frequency using busywait    = 50012
actual frequency using polledTimer = 50200

requested bus clock = 100000
actual frequency using busywait    = 100050
actual frequency using polledTimer = 100806

requested bus clock = 400000
actual frequency using busywait    = 400801
actual frequency using polledTimer = 431034

requested bus clock = 800000
actual frequency using busywait    = 803212
actual frequency using PolledTimer = 657894

Below 100KHz the larger granularity doesn't have as much of an effect, but I've only seen a few people running slow.  Personally, I'd rather have the accuracy above 100KHz and the 800KHz top speed, but it's your call.

Reply:
One other thing I just thought of: I suspect the compiler is putting the busywait() inline, so the shorter routine I did accounts for the 1460 bytes of IRAM saved over the current implementation.  I'll have to go back and retest with the polledTimer to see what that does to the IRAM.  It's a bigger function, so the compiler SHOULD leave it in IRAM and do long calls to it.

Reply:
The polledTimer uses 40 bytes more IRAM and 16 bytes less Flash,

with busywait()

IROM   : 239308          - code in flash
IRAM   : 27596   / 32768 - code in IRAM
DATA   : 1264  )         - initialized variables
RODATA : 888   ) / 81920 - constants
BSS    : 25464 )         - zeroed variables

with polledTimer()

IROM   : 239292          - code in flash
IRAM   : 27636   / 32768 - code in IRAM
DATA   : 1264  )         - initialized variables
RODATA : 888   ) / 81920 - constants
BSS    : 25464 )         - zeroed variables

Reply:
I was wrong on the minimum delay from the polledTimer a little: the minimum is 214ns at 160, and 506ns at 80.  That means the maximum achievable speed is 328KHz @ 80 or 648KHz @ 160, quite short of the 400/800 I got with the modified busywait().

Here's the measured differences for desired vs. actual bus frequency (the numbers a couple of posts back were calculated):
at 80MHz CPU,
50KHz = 52.4
100KHz = 108
400KHz = 328

at 160MHz CPU,
50KHz = 50.74
100KHz = 103
400KHz = 429
800KHz = 648

Reply:
Here's the code difference for the polledTimer:
```cpp
void Twi::setClock(unsigned int freq)
{
    if (freq < 250)  // minimum freq 250Hz
        freq = 250;
    
    preferred_si2c_clock = freq;

#if F_CPU == FCPU80

    if (freq > 300000)
        freq = 300000;
    twi_dcount = (500000000 / freq);  // half-cycle period in ns
    twi_dcount = (twi_dcount - (1120 + 506));  // (half cycle - (overhead + polledTimer minimum) 
    
#else

    if (freq > 630000)
        freq = 630000;
    twi_dcount = (500000000 / freq);  // half-cycle period in ns
    twi_dcount = (twi_dcount - (560 + 214));  // (half cycle - (overhead + polledTimer minimum)

#endif
}


 void ICACHE_RAM_ATTR Twi::busywait(unsigned int v)
{
	esp8266::polledTimeout::oneShotFastNs timeout(v);
	while(!timeout) {}
}
```

Reply:
I think polledTimeout probably needs further optimization and/or specific simplified policies to achieve a similar level of performance as your nop loop. Example: skip some checks. 
That requires further work in polledTimeout itseld, which is outside the scope of this fix. Let's handle that as a separate issue.
Your proposal here is a strict improvement over the current code, despite having the same drift flaw. I'm ok with that.